convert some typeof tags to small integers

vtjnash opened the pull request:
This allows us to avoid needing relocations for some builtin objects (esp. Strings) when loading the system image, saving many objects from being faulted into memory. Takes some of the same code from #48432, but in a different direction so it is applicable to GC, codegen, and pkgimages too.

Closes #48432

vtjnash committed:
change CodeInfo compression from Array to String

String is simpler and smaller (since we do not yet have a Buffer{UInt8}
type, which might be almost as simple) making it easier to manipulate.
Combined with the previous changes, this lets us eliminate many more
relocations from the image (previously every Array->data), which is
quite a few megabytes of array header objects simply gone.

MASTER:
sysimg size breakdown:
     sys data: 79207516
  isbits data: 58603452
      symbols:   463610
    tags list:  1078555
   reloc list:  4491653
    gvar list:   118848
    fptr list:   214000
.text          11109176
.data         144184704

BEFORE (moving existing Strings):
sysimg size breakdown:
     sys data: 75146604
  isbits data: 62679532
      symbols:   463411
    tags list:  1015525
   reloc list:  4481370
    gvar list:   118688
    fptr list:   213952
.text          11129192
.data         144126152

AFTER (making CodeInfo into Strings):
     sys data: 71348124
  isbits data: 63789244
      symbols:   463411
    tags list:   933984
   reloc list:  4398377
    gvar list:   118688
    fptr list:   213904
.text          11132968
.data         141272824

vtjnash committed:
move simple (isbits) objects into the constdata section

vtjnash commented:
@nanosoldier `runbenchmarks("scalar", vs=":master")`

vtjnash commented:
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

JeffBezanson commented on src/julia.h:
```diff
@@ -989,7 +1048,7 @@ STATIC_INLINE jl_value_t *jl_svecset(
 STATIC_INLINE jl_value_t *jl_svecref(void *t JL_PROPAGATES_ROOT, size_t i) JL_NOTSAFEPOINT
 {
-    assert(jl_typeis(t,jl_simplevector_type));
+    assert(jl_typetagis(t,jl_simplevector_tag << 4));
```
Move the `<< 4` into the macro?

JeffBezanson commented on src/julia.h:
```diff
@@ -1055,13 +1114,13 @@ STATIC_INLINE jl_value_t *jl_array_ptr_set(
 {
     assert(i < jl_array_len(a));
-    assert(jl_typeis(a, jl_array_uint8_type));
+    assert(jl_typetagis(a, jl_array_uint8_type));
```
```suggestion
    assert(jl_typeis(a, jl_array_uint8_type));
```

vtjnash replied:
`jl_typetagis` is always faster than `jl_typeis`, which is kept as a legacy alias, if you can guarantee that the type is not sometimes represented as a tag.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/49556#issuecomment-1531957537) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/c2096e9_vs_6052e44/report.md).

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/49556#issuecomment-1531957917) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/2910899_vs_6052e44/report.md).

vtjnash commented:
@nanosoldier `runbenchmarks("union", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/49556#issuecomment-1532412417) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/9907de7_vs_e9d678e/report.md).

vtjnash commented:
Fascinating. I didn’t expect this to improve performance by that much, but seems consistent in the two runs.

gbaraldi commented:
I wonder if removing the load makes it codegen a bit better.

vtjnash commented:
Okay, it turns out our codegen for this is just pretty bad:
```
    %22 = (%19 === Main.nothing)::Bool
```
```
;  @ REPL[1]:4 within `perf_countnothing`
  %94 = and i8 %tindex_phi3, 127, !dbg !67
  %95 = icmp eq i8 %94, 1, !dbg !67
  %96 = select i1 %95, {} addrspace(10)* addrspacecast ({}* inttoptr (i64 4815131216 to {}*) to {} addrspace(10)*), {} addrspace(10)* null, !dbg !67
  %97 = icmp eq i8 %94, 2, !dbg !67
  %98 = select i1 %97, {} addrspace(10)* addrspacecast ({}* inttoptr (i64 4815126496 to {}*) to {} addrspace(10)*), {} addrspace(10)* %96, !dbg !67
  %99 = icmp eq {} addrspace(10)* %98, addrspacecast ({}* inttoptr (i64 4815131216 to {}*) to {} addrspace(10)*), !dbg !67
  %100 = zext i1 %99 to i8, !dbg !67
```
But this PR, as a side-effect, simplified that slightly, which let LLVM mostly optimize it away in this case:
```
;  @ REPL[1]:4 within `perf_countnothing`
  %94 = and i8 %tindex_phi3, 127, !dbg !67
  %95 = icmp eq i8 %94, 1, !dbg !67
  %96 = select i1 %95, i64 4847026336, i64 0, !dbg !67
  %97 = icmp eq i8 %94, 2, !dbg !67
  %98 = select i1 %97, i64 4847021616, i64 %96, !dbg !67
  %99 = icmp eq i64 %98, 4847026336, !dbg !67
  %100 = zext i1 %99 to i8, !dbg !67
```
Vs. if we replace that with `x isa Nothing`, which emits the optimal code directly:
```
  %94 = and i8 %tindex_phi3, 127, !dbg !67
  %95 = icmp eq i8 %94, 1, !dbg !67
  %96 = zext i1 %95 to i8, !dbg !67
```
```
function perf_countnothing(X::AbstractArray); n = 0; @inbounds for x in X; n += x === nothing; end; n; end
@code_llvm raw=true optimize=false debuginfo=:source perf_countnothing(Vector{Union{Float64, Nothing}}(undef,0))
```

vtjnash commented:
The fix for that conflicts slightly with our change here, because we need to edit some of the lines that need to be moved, so I can do that as a followup after merging this, if it is ready to go, or before if this ends up needing to be rebased anyways.

vtjnash commented:
`isa` is special cased, but `===` was not falling back to it correctly

vtjnash committed:
convert some typeof tags to small integers

The value here is two-fold. One, we speed up sysimg load times slightly,
since we can entirely skip relocations for these values and avoid
faulting in some writable pages. This should let us (later) move more
content into the ConstData section, such as String objects. Secondly,
some type-manipulation native code becomes simpler, since it is based on
small consecutive constants instead of requiring extra pointer loads to
check for pointer identity. This is similar to the existing small-union
optimization, but for type tags, and only for specific ones.

This makes use of the fact that the zero page (about 4096 byes) cannot
be allocated in the usual C memory model, which lets us define up to 255
of these special types, after taking into account the 4 gc bits. As a
current implementation limitation for performance, these cannot be
parametric types.

Note there are probably more places in Base that can be updated to use
`jl_typetagof` instead of `jl_typeof`, where we know if it is a type or tag.
For example, this optimize most of builtins.c file, except for the
`jl_object_id` function.
